doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1446 @@
+
+2.1.13 Dynamic and Static NAT/NAPT entries should persist across warm reboot with no traffic disruption to the active flows.
+2.1.14 NAT feature can be enabled or disabled globally. All the NAT configuration and functionality works only if the NAT feature is enabled.
+
+2.2.0 Provide the ability for Twice NAT that translates both Source and Destination IP addresses when crossing the zones.
+
+2.2.1 Provide the ability for Twice NAPT that translates both Source and Destination IP addresses and L4 ports when crossing the zones.
+
+2.3.0 ICMP packets are translated via NAT or NAPT rules in the Linux kernel.
+
+2.4.0 The hardware NAT table full condition is handled gracefully by logging error log message.
+
+2.5.0 The SNAT or DNAT miss packets are rate limited to CPU.
+
+2.5.1 The NAT miss packets are processed from a higher priority CPU COS Queue than the Broadcast/Unknown Multicast packets.
+
+2.6.0 Support for the Full Cone NAT/NAPT functionality.
+
+3.0 ACL and NAT pool binding is applicable on the valid L3 interfaces (VLAN, Ethernet, PortChannel) defined by the ACL TABLE entry.
+
+3.1 The L3 ports on which the NAT ACL is applied are in a different NAT zone compared to the port on which the NAT Pool IP address is based on.
+
+3.2 Provide configuration of zones on L3 interfaces.
+
+4.0 Provide configurable age timeout interval for the inactive dynamic UDP NAPT entries (in seconds). Default is 300 secs. Range is from 120 sec to 600 secs.
+
+4.1 Provide configurable age timeout interval for the inactive dynamic TCP NAPT entries (in seconds). Default is 86400 secs. Range is 300 sec to 432000 secs (5 days).
+
+4.2 Provide configurable age timeout interval for the inactive dynamic NAT entries (in seconds). Default is 600 secs. Range is 300 sec to 432000 secs (5 days).
+
+5.0 Should be able to ping from internal host to an outside host via NAPT.
+
+5.1 Should be able to traceroute from internal host to an outside host via NAPT.
+
+5.2 Provide support for translation statistics per zone and per NAT flow.
+
+5.3 Ability to clear the NAT translation table entries.
+
+5.4 Ability to clear the NAT translation statistics.
+
+5.2 Ability to stop and start the NAT docker service.
+
+5.3 Ability to enable the logging at different severity levels of the NAT module.
+
+## 1.2 Configuration and Management Requirements
+Configuration of the NAT feature can be done via:
+
+- JSON config input
+- Incremental CLI
+
+## 1.3 Scalability Requirements
+Ability to support the scale of up to 40K Bi-directional NAT entries, but this is SAI capability dependent.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+NAT feature is targeted for DC and Enterprise deployments.
+
+In DC deployments, the ToR switch can enable NAT feature to translate the traffic from/to the attached Servers running in the private network and the public domain.
+![DC deployment use case](images/nat_dc_deployment_usecase.png)
+
```

Comment:
i feel we need to explain a little bit more on the DC deployment use case here. Can you add these three packet flows in the diagram. Then, the design can centered around how to support these three scenarios.

1. to and from internet, we do the DNAT and SNAT.
2. within DC, we do not need to do the NAT. Can use private IP for communication, save ASIC resource.
3. within the TOR, if server needs to talks to each other. In some scenarios, we need to DNAT and SNAT via the TOR.

Reply:
Done

Reply:
Thanks.  I see that the test case " Modify the ACL rules and verify that the iptable rules are updated in the kernel. " is added.  It seems corresponding conntrack or hardware rules may not be cleaned up after ACL rules change?  If so, please mark it as the current limitation.

Reply:
Mentioned as a limitation in section 3.3.1.1